### PR TITLE
Shladkyi - refactor to Selenide

### PR DESCRIPTION
**Pull request by SerhiiHladkyi**

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java [hunk 1]
```diff
@@ -16,25 +17,21 @@ public GoogleHomePage open() {
...
-        TestRunner.getWait().until(ExpectedConditions.elementToBeClickable(searchSettingsButton));
-        searchSettingsButton.click();
+        WebElementUtil.clickElement("//button[@id='Mses6b']");
+        WebElementUtil.clickElement("//a[@class='EzVRq']");
         return new GoogleSettingsPage();
     }
 
     public String getSearchButtonText() {
-        return TestRunner.getDriver().findElement(By.cssSelector("[name='btnK']")).getAttribute("value");
+        return TestRunner.getDriver().findElement(By.xpath("//input[@name='btnK']")).getAttribute("value");
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java]: use wait

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java [hunk 2]
```diff
@@ -16,25 +17,21 @@ public GoogleHomePage open() {
...
     }
 
     public String getLuckyButtonText() {
-        return TestRunner.getDriver().findElement(By.cssSelector("[name='btnI']")).getAttribute("value");
+        return TestRunner.getDriver().findElement(By.xpath("//input[@name='btnI']")).getAttribute("value");
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java]: use wait

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java (hunk 1 above)
Comment (SerhiiHladkyi) [reply src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java]: When i use wait on this element I always have TimeoutException. I cant find why this works like this.

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java (hunk 2 above)
Comment (SerhiiHladkyi) [reply src/main/java/com/softserveinc/ita/shladkyi/GoogleHomePage.java]: When i use wait on this element I always have TimeoutException. I cant find why this works like this.

File: src/main/java/com/softserveinc/ita/common/WebElementUtil.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    public static void setValueForElement(String webElementXPath, String value) {
+        TestRunner.getWait().until(ExpectedConditions.elementToBeClickable(By.xpath(webElementXPath))).sendKeys(value);
+    }
+
+    public static List<WebElement> getListOfElements(String webElementsXPath) {
+        return TestRunner.getWait().until(ExpectedConditions.visibilityOfAllElementsLocatedBy(By.xpath(webElementsXPath)));
+    }
+
+    public static WebElement getElementFromListByIndex(String webElementsXPath, int elementIndex) {
+        return getListOfElements(webElementsXPath).get(elementIndex);
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/common/WebElementUtil.java]: still inefficient

File: src/main/java/com/softserveinc/ita/common/WebElementUtil.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public static List<WebElement> getListOfElementsMoreThenAmount(String webElementXpath, int amount) {
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/common/WebElementUtil.java]: rename

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleSearchResultPage.java [hunk 5]
```diff
@@ -1,25 +1,26 @@
...
 import org.openqa.selenium.WebElement;
 
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class GoogleSearchResultPage {
     public String getSearchResultLinkText(int indexOfLink) {
-        List<WebElement> allSearchResultLinks = TestRunner.getDriver().findElements(By.xpath("//h3[@class='LC20lb DKV0Md']"));
-        return allSearchResultLinks.get(indexOfLink).getText();
+        return WebElementUtil.getElementFromListByIndex("//h3[@class='LC20lb DKV0Md']", indexOfLink)
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/shladkyi/GoogleSearchResultPage.java]: inefficient

File: src/main/java/com/softserveinc/ita/vpetrat/GoogleHomePage.java [hunk 6]
```diff
@@ -13,22 +12,13 @@ public GoogleHomePage openHomePage() {
...
-        TestRunner
-                .getWait()
-                .until(ExpectedConditions.elementToBeClickable(By.xpath("//input[@class='gLFyf gsfi']")))
-                .sendKeys(request);
-        TestRunner
-                .getWait()
-                .until(ExpectedConditions.visibilityOfAllElementsLocatedBy(By.xpath("//input[@name='btnI']")))
-                .get(0).click();
+        WebElementUtil.setElementValue("//input[@class='gLFyf gsfi']", request);
+        WebElementUtil.getElementFromListByIndex("//input[@name='btnI']", 0).click();
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/vpetrat/GoogleHomePage.java]: inefficient

File: src/main/java/com/softserveinc/ita/vpetrat/GoogleSearchResultsPage.java [hunk 7]
```diff
@@ -33,11 +27,7 @@ public String getSearchResultLinkTextForIndex(int index) {
     }
 
     public GoogleSearchImagesPage navigateToImagesPage() {
-        TestRunner
-                .getWait()
-                .until(ExpectedConditions.visibilityOfAllElementsLocatedBy(By.xpath("//a[@class='hide-focus-ring']")))
-                .get(0)
-                .click();
+        WebElementUtil.getElementFromListByIndex("//a[@class='hide-focus-ring']", 0).click();
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/vpetrat/GoogleSearchResultsPage.java]: same

**Review by toleksyn — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/shladkyi/GoogleSearchResultPage.java [hunk 8]
```diff
@@ -1,25 +1,28 @@
...
+        return $x(String.format("(%s)[%d]", "//h3[@class='LC20lb DKV0Md']", index + 1)).getText();
     }
 
     public List<String> getAllSearchResultLinks() {
-        List<WebElement> allSearchResultLinks = TestRunner.getDriver().findElements(By.xpath("//div[@class='TbwUpd NJjxre']"));
-        return allSearchResultLinks.stream().map(WebElement::getText).collect(Collectors.toList());
+        return $$x("//div[@class='TbwUpd NJjxre']")
+                .stream()
+                .map(SelenideElement::getText)
+                .collect(Collectors.toList());
```
Comment (toleksyn) [inline src/main/java/com/softserveinc/ita/shladkyi/GoogleSearchResultPage.java]: $$x().texts()

File: src/test/java/com/softserveinc/ita/pageobjects_task/shladkyi/GoogleTest.java [hunk 9]
```diff
@@ -35,13 +38,13 @@ public void testValidGoogleSearchAnyLink() {
...
         GoogleSearchImagesResultPage googleImagesResultPage = googleHomePage.searchFor("funny kitten").openImagesPage();
-        List<String> imagesListText = googleImagesResultPage.getSearchResultImagesTextList();
+        List<String> imagesListText = googleImagesResultPage.getSearchResultImagesTextList(10);
         Assert.assertTrue(imagesListText.size() >= 10);
         String testText = "funny";
         Assert.assertTrue(imagesListText.get(0).toLowerCase().contains(testText));
         Assert.assertTrue(imagesListText.get(4).toLowerCase().contains(testText));
         googleImagesResultPage.openGoogleHomePage();
-        Assert.assertTrue(TestRunner.getDriver().getCurrentUrl().contains("https://www.google.com/"));
+        Assert.assertTrue(requireNonNull(title()).contains("Google"));
```
Comment (toleksyn) [inline src/test/java/com/softserveinc/ita/pageobjects_task/shladkyi/GoogleTest.java]: requireNonNull - why is this needed?

**Review by toleksyn — CHANGES_REQUESTED** [review]
